src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+    private static final int DEFAULT_DIVISOR = 2;
+    private static final int INDEX_STOPPING_FROM_TAIL = 1;
+    private static final int FIRST_ELEMENT_INDEX = 0;
+    private static final int EMPTY_LIST_VALUE = 0;
+    private Node<T> head;
+    private Node<T> tail;
+    private Node<T> current;
+    private int size;
+
     @Override
     public void add(T value) {
+        current = new Node<>(null, value, null);
+        if (tail == null) {
+            tail = current;
+            head = tail;
+        } else {
+            current.prev = tail;
+            current.prev.next = current;
+            tail = current;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddIndexExistence(index);
+        if (index == FIRST_ELEMENT_INDEX) {
+            head = new Node<>(null, value, head);
+            tail = head;
+        } else if (index == size) {
+            add(value);
+            return;
+        } else {
+            current = getNodeByIndex(index);
+            current = new Node<>(current.prev, value, current);
+            current.next.prev = current;
+            current.prev.next = current;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        current = getNodeByIndex(index);
+        T oldObject = current.item;
+        current.item = value;
+        return oldObject;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        current = getNodeByIndex(index);
+        if (current.next == null && current.prev == null) {
+            head = null;
+            tail = null;
+        } else if (current.next == null) {
+            current.prev.next = null;
+            tail = current.prev;
+        } else if (current.prev == null) {
+            current.next.prev = null;
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
+        current = head;
+        for (int i = 0; i < size; i++) {
+            if ((object == current.item)
+                    || (object != null && object.equals(current.item))) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == EMPTY_LIST_VALUE;
+    }
+
+    private static class Node<T> {
+        private T item;
```

Comment:
```suggestion
        private T value;
```
It would be better, i think.